test/OpenTelemetry.Tests/Trace/Propagation/BaggagePropagatorTest.cs

```diff
@@ -159,5 +185,23 @@ public void ValidateBaggageInjection()
             Assert.Single(carrier);
             Assert.Equal("key1=value1,key2=value2", carrier[BaggagePropagator.BaggageHeaderName]);
         }
+
+        [Fact]
+        public void ValidateSpecialCharsBaggageInjection()
+        {
+            var carrier = new Dictionary<string, string>();
+            var propagationContext = new PropagationContext(
+                default,
+                new Baggage(new Dictionary<string, string>
+                {
+                    { "key 1", "value 1" },
+                    { "key2", "!x_x,x-x&x(x\");:" },
+                }));
+
+            this.baggage.Inject(propagationContext, carrier, Setter);
+
+            Assert.Single(carrier);
+            Assert.Equal("key%201=value%201,key2=%21x_x%2Cx-x%26x%28x%22%29%3B%3A", carrier[BaggagePropagator.BaggageHeaderName]);
```

Comment:
I think `!` can be optimized by just using one single character `!` instead of `%21` here.
Here goes the spec https://github.com/w3c/baggage/blob/master/baggage/HTTP_HEADER_FORMAT.md#definition.

Reply:
The same for `_` vs. `%26`, `&` vs. `%28` and `(` vs. `%29`.

Reply:
I think we want to have the implementation which generates the shortest representation given the W3C Baggage does put limitation on the size.

Reply:
This approach produces an exception. You can have a look at issue #1607 /
It looks like:
```
[2020-11-22 15:23:48Z ERR] Connection id "0HM4EVAE30FTQ", Request id "0HM4EVAE30FTQ:00000001": 
An unhandled exception was thrown by the application. System.FormatException: The format of
value '!x_x%2Cx-x%26x(x' is invalid.
   at System.Net.Http.Headers.NameValueHeaderValue.CheckValueFormat(String value)
   at System.Net.Http.DiagnosticsHandler.InjectHeaders(Activity currentActivity, HttpRequestMessage request)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at FrontendService.Controllers.WeatherFrontendController.GetWeatherForecast() in D:xxxxxxxxxxxxx:line 109
```

`System.Net.Http.Headers.NameValueHeaderValue.CheckValueFormat` throws an exception for such symbols. That's why I've made it safer. Another approach — to replace this check with something else. But it could be done by another PR.

The main purpose of this PR — make this code workable without serious performance degradation. At the moment the user can't use all of these symbols at all.

Reply:
I think I got what you want. It's a good point. :)

Let me rephrase your sentence to ensure I got you right.
Am I right that you want the following flow:
1. We have raw baggage that was added like `!x_x,x-x&x(x`
2. Then we convert it with `WebUtility.UrlEncode` so the value will be transformed into `!x_x%2Cx-x%26x(x`
3. Then we send it to the recipient app
4. Then the recipient app decodes it with `WebUtility.UrlDecode` and got the raw value
5. But at this moment we have to call `NameValueHeaderValue.TryParse` and for this purpose, we encode the value again with `Uri.EscapeDataString` and then use `Uri.UnescapeDataString` after parsing

Is this correct? :)

Reply:
I've made the changes. Please take a look :)

Reply:
The flow looks reasonable, there are room to improve the perf and memory use (e.g. reduce heap allocation) though (which I guess is not a blocker for this PR).

Reply:
Agree, the performance could be improved in the separate PR.